language/move-prover/spec-lang/src/translate.rs

```diff
@@ -1446,21 +1463,63 @@ impl<'env, 'translator> ModuleTranslator<'env, 'translator> {
             }
             let translated = et.translate_seq(&loc, &seq, &result_type);
             et.finalize_types();
-            // TODO(emmazzz): right now we can't detect a pure-looking move
-            // function which calls an impure move function. Later we need
-            // to come up with more general algorithm for detecting impure
-            // move functions.
             // If no errors were generated, then the function is considered pure.
             if !*et.errors_generated.borrow() {
+                et.called_spec_funs.iter().for_each(|(mid, fid)| {
+                    self.parent.env.add_edge_to_call_graph(
+                        self.module_id,
+                        SpecFunId::new(fun_idx),
+                        *mid,
+                        *fid,
+                    );
+                });
                 self.spec_funs[self.spec_fun_index].body = Some(translated);
-                self.parent
-                    .fun_table
-                    .entry(full_name)
-                    .and_modify(|e| e.is_pure = true);
             }
         }
         self.spec_fun_index += 1;
     }
+
+    /// Propagate the impurity of move functions from callees to callers so
+    /// that we can detect pure-looking move functions which calls impure
+    /// move functions.
+    fn propagate_function_impurity(
+        &mut self,
+        mut visited: &mut BTreeMap<usize, bool>,
```

Comment:
Instead of using usize, using the id type (whatever it is, SpecFunId or FunId) would help to make the code better understandable.